src/main/java/com/google/devtools/build/lib/actions/cache/VirtualActionInput.java

```diff
@@ -48,7 +54,11 @@ default FileArtifactValue getMetadata() throws IOException {
    * use instances of this class to represent those files.
    */
   final class EmptyActionInput implements VirtualActionInput {
-    private final PathFragment execPath;
+    @Nullable private final PathFragment execPath;
+
+    private EmptyActionInput() {
+      execPath = null;
```

Comment:
Instead of setting the field to null and handling that case in all of the methods, it would be easier to just have a private class for the empty marker singleton.

Reply:
Yes, that sounds better. I didn't know if this was the only use of `EmptyActionInput` or if some Google code makes real execpath empty inputs.

Reply:
There is one internal use of `EmptyActionInput` that also passes `/dev/null`. I suggest keeping this class as it was for now and just create a new `VirtualActionInput` implementation class for the "null" case. I will separately see if the internal use case can be migrated similarly and if so follow up with a change to delete this class.

Reply:
Okay, reverting to the original version of the PR now...